### PR TITLE
fix(dot/sync): fix "block with unknown header is ready" error

### LINKED ISSUE #2189: fix "block with unknown header is ready" error
## Task summary

- when syncing on the devnet, the error "block with unknown header is ready" shows up somewhat often
- this really shouldn't be happening, as this means the `handleReadyBlock` function is being called with a block that isn't ready (ie. it's header isn't available to us)
- the error is probably coming from `tipSyncer.handleTick`, as when `handleReadyBlock` is called from `doSync`, the block data is verified to already have headers

### PR DESCRIPTION
**Pull request by noot**
## Changes

- fix "block with unknown header is ready" error which was caused by already having the block in the ready queue or processed, so it was deleted from the pending set already (which is why we couldn't find the header in the pending set)

## Tests

```
go test ./dot/sync
```

## Issues

- closes #2189

## Primary Reviewer

- @EclesioMeloJunior

### CODE REVIEW

**Review by jimjbrettj — APPROVED** [review]
nice work!

File: dot/sync/chain_sync.go [hunk 1]
```diff
@@ -692,41 +693,61 @@ func (cs *chainSync) doSync(req *network.BlockRequestMessage, peersTried map[pee
...
+			// block wasn't in the pending set!
+			// let's check the db as maybe we already processed it
+			has, err := cs.blockState.HasHeader(bd.Hash)
+			if err != nil && !errors.Is(err, chaindb.ErrKeyNotFound) {
+				logger.Debugf("failed to check if header is known for hash %s: err=%s", bd.Hash, err)
+				return
+			}
+
+			if has {
+				// seems we already processed the block, so let's just return
```
Comment (qdm12) [inline dot/sync/chain_sync.go:723]: nit unneded the log message says it already

```suggestion
```

File: dot/sync/chain_sync.go [hunk 2]
```diff
@@ -692,41 +693,61 @@ func (cs *chainSync) doSync(req *network.BlockRequestMessage, peersTried map[pee
...
 	}
 
 	return nil
 }
 
-func handleReadyBlock(bd *types.BlockData, pendingBlocks DisjointBlockSet, readyBlocks *blockQueue) {
-	// see if there are any descendents in the pending queue that are now ready to be processed,
-	// as we have just become aware of their parent block
+func (cs *chainSync) handleReadyBlock(bd *types.BlockData) {
+	// first check that the block isn't already in the ready queue
```
Comment (qdm12) [inline dot/sync/chain_sync.go:703]: nit unneeded

```suggestion
```

File: dot/sync/chain_sync.go [hunk 3]
```diff
@@ -692,41 +693,61 @@ func (cs *chainSync) doSync(req *network.BlockRequestMessage, peersTried map[pee
...
+			}
+
+			if has {
+				// seems we already processed the block, so let's just return
+				logger.Tracef("ignoring block we've already processed, hash=%s", bd.Hash)
+				return
+			}
+
+			// this is bad and shouldn't happen
+			logger.Errorf("block with unknown header is ready: hash=%s", bd.Hash)
```
Comment (qdm12) [inline dot/sync/chain_sync.go:729]: Would this happen if `errors.Is(err, chaindb.ErrKeyNotFound)`? :thinking:

File: dot/sync/tip_syncer.go [hunk 4]
```diff
@@ -8,23 +8,29 @@ import (
 	"math/big"
 
 	"github.com/ChainSafe/gossamer/dot/network"
+	"github.com/ChainSafe/gossamer/dot/types"
 	"github.com/ChainSafe/gossamer/lib/common"
 )
 
 var _ workHandler = &tipSyncer{}
 
+type handleReadyBlockFunc = func(*types.BlockData)
```
Comment (qdm12) [inline dot/sync/tip_syncer.go:17]: nit does the following work?

```suggestion
type handleReadyBlockFunc func(*types.BlockData)
```

**Review by qdm12 — APPROVED** [review]

File: dot/sync/chain_sync.go [hunk 5]
```diff
@@ -692,41 +693,61 @@ func (cs *chainSync) doSync(req *network.BlockRequestMessage, peersTried map[pee
...
 	if bd.Header == nil {
-		block := pendingBlocks.getBlock(bd.Hash)
+		block := cs.pendingBlocks.getBlock(bd.Hash)
 		if block == nil {
-			logger.Criticalf("block with unknown header is ready: hash=%s", bd.Hash)
+			// block wasn't in the pending set!
+			// let's check the db as maybe we already processed it
+			has, err := cs.blockState.HasHeader(bd.Hash)
+			if err != nil && !errors.Is(err, chaindb.ErrKeyNotFound) {
+				logger.Debugf("failed to check if header is known for hash %s: err=%s", bd.Hash, err)
```
Comment (EclesioMeloJunior) [inline dot/sync/chain_sync.go:718]:
```suggestion
				logger.Debugf("failed to check if header is known for hash %s: %s", bd.Hash, err)
```

File: dot/sync/chain_sync.go (hunk 3 above)
Comment (EclesioMeloJunior) [reply dot/sync/chain_sync.go:729]: this code already checks for `!errors.Is(err, chaindb.ErrKeyNotFound)` so if `has` is false then we don't need to check for this `chaindb.ErrKeyNotFound` error

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: dot/sync/tip_syncer.go (hunk 4 above)
Comment (noot) [reply dot/sync/tip_syncer.go:17]: yes it does, is there a preference for one?

File: dot/sync/chain_sync.go (hunk 5 above)
Comment (noot) [reply dot/sync/chain_sync.go]: done!

**Review by EclesioMeloJunior — APPROVED** [review]

**Pull request merged**